modules/oe_translation_poetry/src/Poetry.php

```diff
@@ -41,7 +94,167 @@ public function __construct(ConfigFactoryInterface $configFactory, LoggerChannel
+    }
+
     parent::__construct($values);
+
+    // Register our own service provider to override the SOAP client. We need
+    // to pass the current cookies to that the SOAP client uses them in its
+    // requests. This is critical for ensuring the tests work.
+    $this->register(new PoetrySoapProvider($requestStack->getCurrentRequest()->cookies->all()));
+
+    $this->translatorSettings = $settings;
+    $this->state = $state;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->database = $database;
+  }
+
+  /**
+   * Gets the global identification number.
+   *
+   * @return string|null
+   *   The number.
+   */
+  public function getGlobalIdentifierNumber(): ?string {
+    return $this->state->get('oe_translation_poetry_id_number');
+  }
+
+  /**
+   * Sets the global identification number.
+   *
+   * @param string $number
+   *   The number.
+   */
+  public function setGlobalIdentifierNumber(string $number): void {
+    $this->state->set('oe_translation_poetry_id_number', $number);
+  }
+
+  /**
+   * Returns the identifier for making a translation request for a content.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The entity.
+   *
+   * @return \EC\Poetry\Messages\Components\Identifier
+   *   The identifier.
+   */
+  public function getIdentifierForContent(ContentEntityInterface $entity): Identifier {
+    $last_identifier_for_content = $this->getLastIdentifierForContent($entity);
+    if ($last_identifier_for_content instanceof Identifier) {
+      // If the content has already been translated, we need to use the
+      // identifier it had and just increase the version. This means potentially
+      // even an older number than the most current global one as well as an
+      // older year.
+      $last_identifier_for_content->setVersion($last_identifier_for_content->getVersion() + 1);
+      return $last_identifier_for_content;
+    }
+
+    $identifier = $this->getIdentifier();
+    $number = $this->getGlobalIdentifierNumber();
+
+    if (!$number) {
+      // If we don't have a number it means it's the first ever request.
+      $identifier->setSequence(Settings::get('poetry.identifier.sequence'));
+      return $identifier;
+    }
+
+    // If we have a global number, we can maybe use it. However, we first to
+    // determine the part. And for this we need to check the jobs.
+    $part = $this->getLastPartForNumber($number);
+    if ($part > -1) {
+      // We check if the part came back as -1 in case jobs were missing from
+      // the system, we increment only if we know where to increment from.
+      $part++;
+    }
+
+    // If the incremented part is 100, we need to scrap the the global number
+    // and request a new one. The maximum can be 99.
+    if ($part === 100) {
+      $identifier->setSequence(Settings::get('poetry.identifier.sequence'));
+      return $identifier;
+    }
+
+    $identifier->setNumber($number);
+
+    if ($part > -1) {
```

Comment:
If I follow the code correctly, if the getLastPartForNumber ever returns a -1 we will set the part to be 0 (as defined in the constructor). However, if we ever get a -1 I believe we should behave differently since this would mean we have lost track of the part for the particular number we are checking for. In this case we should either throw and exception or request a new number right?

Reply:
How about now? If we get -1, we set the sequence to request a new number.

Reply:
I think this is a solution that works from our side but will not be very liked in the DGT side. In any case, I don't think we have any better solution for now.